src/mono/System.Private.CoreLib/src/System/Reflection/Emit/ModuleBuilder.Mono.cs

```diff
@@ -343,10 +343,10 @@ public MethodInfo GetArrayMethod(Type arrayClass, string methodName, CallingConv
         public EnumBuilder DefineEnum(string name, TypeAttributes visibility, Type underlyingType)
         {
             ITypeIdentifier ident = TypeIdentifiers.FromInternal(name);
-            if (name_cache.ContainsKey(ident))
+            if (name != null && name_cache.ContainsKey(ident))
```

Comment:
I suspect that we will likely see more instances of breaks like this one in the code out there.

Do we need to file a breaking change notification for this change?

Reply:
So you think we should file a breaking change notification to indicate that the `GetHashCode` method will be called on a key provided to an initially empty `Dictionary` whereas before it did not?

Reply:
I would like to hear what others think about this issue.

Reply:
cc @marklio @GrabYourPitchforks for thoughts.

Reply:
It's not clear to me what the break was. Was there a bug in the implementation of `<some_concrete_type>.GetHashCode` where it was throwing an exception when called? Normally one would expect this method to be called as part of dictionary / hashset processing.

Reply:
```csharp
using System;
using System.Collections.Generic;

var d = new Dictionary<object, string>();
Console.WriteLine(d.ContainsKey(new Sentinel()));

class Sentinel
{
    public override int GetHashCode()
    {
        throw new NotImplementedException();
    }
}
```
Before this change: works fine
After this change: fails with `NotImplementedException`

The behavior change can also manifest itself as performance regression if the code expects lookup in empty dictionary to be very fast. For example, replace `throw new NotImplementedException()` in my example with some expensive computation.

Reply:
IMO the behavioral change seems like the kind of thing we'd take in a major version. But getting this change in early would help us flesh out just how large a break it would be. We can always revert if it ends up harming the ecosystem.

I understand the perf regression scenario, but wouldn't we expect the majority of dictionaries to be non-empty? That seems like it would be a perf improvement in the common case, with only a few call sites seeing a regression.

Reply:
@marklio If I remember correctly, we have done analysis data on the capacity distribution for common collections in real apps. How common are empty dictionaries?

Reply:
Also, it would take just one extra if-check to mitigate this issue: https://github.com/dotnet/runtime/pull/48725/files#r593419736

Reply:
@TylerBrinkley do you know how reintroducing a check here will affect perf results? I'm guessing that an "is empty" check will cost about the same as the original "is null" check, but it'd be good to confirm.

Reply:
I just saw this thread. I'll see if we had information on empty dictionaries today.

Reply:
The last analysis we did of dictionaries was back in 2018, where we were interested in the usage of types as keys to help inform what investments we should make in dictionaries. Data retention policies mean that the underlying study data is long gone, so I can't provide a bunch of backing data, but from summaries, it did indicate that empty dictionaries are common (although we can't tell what operations/frequency are performed on them, or how long they stay empty). There aren't resources to do another study at this time. My guess based on available data here is that operations on empty dictionaries are pretty common. IMO, I tend to agree with @GrabYourPitchforks that perf-sensitive operations on "chronically empty" dictionaries probably aren't common if they were important enough to measure, and profiling would likely shine a pretty direct light on the culprit if it was important enough to measure (and differing perf characteristics are pretty expected in new versions). The functional breaks associated with this change (which we know exist due to the presence of such fixes in the PR) are likely unexpected and pretty hard to reason about if people hit them, and they may not be in a position to be able to address them if they don't own the types involved. I'd say if an easy acceptable mitigation exists, we should take it.